noise.zig

```diff
@@ -0,0 +1,668 @@
+            try self.cs2.read_msg_regular(msg);
+        } else {
+            try self.cs1.read_msg_regular(msg);
+        }
+
+        self.mc += 1;
+    }
+};
+
+pub const HandshakeState = struct {
+    ss: SymmetricState,
+
+    /// The local static keypair
+    s: ?Ed25519.KeyPair,
+
+    /// The local ephemeral key pair
+    e: ?Ed25519.KeyPair,
+
+    /// The remote parties static public key
+    rs: [KEY_LEN]u8,
+
+    /// The remote parties ephemeral public key
+    re: [KEY_LEN]u8,
+
+    /// Indicates the initiator or responder role.
+    initiator: bool,
+
+    /// Pre-shared Secret Key.
+    psk: [KEY_LEN]u8,
+
+    pub fn clear(self: *HandshakeState) void {
+        self.s = null;
+        self.e = null;
+        self.re = EMPTY_KEY;
+        self.psk = EMPTY_KEY;
+    }
+
+    pub fn init(
+        initiator: bool,
+        prologue: []const u8,
+        s: Ed25519.KeyPair,
+        psk: [KEY_LEN]u8,
+    ) HandshakeState {
+        const protocol_name = "Noise_NX_25519_ChaChaPoly_BLAKE2s";
+        var ss = SymmetricState.initialize_symmetric(protocol_name);
+        ss.mix_hash(prologue);
+
+        return .{
+            .ss = ss,
+            .s = s,
+            .e = null,
+            .rs = EMPTY_KEY,
+            .re = EMPTY_KEY,
+            .initiator = initiator,
+            .psk = psk,
+        };
+    }
+
+    pub fn write_msg_a(self: *HandshakeState, msg: []u8) !void {
+        if (msg.len < KEY_LEN) return error.InvalidMsgLen;
+        if (self.e == null) self.e = try Ed25519.KeyPair.create(null);
+
+        var ne = msg[0..KEY_LEN];
+        mem.copy(u8, ne, &self.e.?.public_key);
+        self.ss.mix_hash(ne);
+
+        self.ss.mix_hash(msg[KEY_LEN..]);
+    }
+
+    pub const CipherStateResult = struct {
+        hash: [HASH_LEN]u8,
+        cs1: CipherState,
+        cs2: CipherState,
+    };
+
+    pub fn write_msg_b(self: *HandshakeState, msg: []u8) !CipherStateResult {
+        if (msg.len < KEY_LEN) return error.InvalidMsgLen;
+        if (self.e == null) self.e = try Ed25519.KeyPair.create(null);
+
+        var ne = msg[0..KEY_LEN];
+        var in_out = msg[KEY_LEN..];
+        mem.copy(u8, ne, &self.e.?.public_key);
+        self.ss.mix_hash(ne);
+
+        self.ss.mix_key(&try dh(self.e.?, self.re));
+
+        var ns = in_out[0 .. KEY_LEN + MAC_LEN];
+        var in_out_2 = in_out[KEY_LEN + MAC_LEN ..];
+        mem.copy(u8, ns, &self.s.?.public_key);
+        try self.ss.encrypt_and_hash(ns);
+
+        self.ss.mix_key(&try dh(self.s.?, self.re));
+        try self.ss.encrypt_and_hash(in_out_2);
+
+        const cs = self.ss.split();
+        self.ss.clear();
+
+        return CipherStateResult{
+            .hash = self.ss.h,
+            .cs1 = cs[0],
+            .cs2 = cs[1],
+        };
+    }
+
+    pub fn read_msg_b(self: *HandshakeState, msg: []u8) !CipherStateResult {
+        if (msg.len < MAC_LEN + KEY_LEN) return error.InvalidMsgLen;
+
+        var re = msg[0..KEY_LEN];
```

Comment:
To improve readability, it would be good to avoid defining slices of arrays as their own separate variables, and directly inline them into the function call's parameters instead (unless the slice's start/end bounds are computed from other variables/constants).